### PR TITLE
[CI]Add Sonar Check

### PR DESCRIPTION
**Pull request by CalvinKirs**
It can help us better check some possible problems with the project
such as #1655

### CODE REVIEW

**Comment by CalvinKirs** [comment]
I'll submit the ticket first.

File: .github/workflows/code-analysys.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: SonarCloud
+on:
+  push:
+    branches:
+      - master
```
Comment (asdf2014) [inline .github/workflows/code-analysys.yml]: Wrong branch..

**Review by asdf2014 — CHANGES_REQUESTED** [review]

File: .github/workflows/code-analysys.yml [hunk 2]
```diff
@@ -24,7 +24,7 @@ on:
 jobs:
   build:
     runs-on: ubuntu-latest
-    if: github.repository == 'apache/incubator-seatunnel'
+    if: null != ${{ secrets.SONARCLOUD_TOKEN }}
```
Comment (zhongjiajie) [inline .github/workflows/code-analysys.yml]: Dose `if: github.repository == 'apache/incubator-seatunnel'` not work?

Comment (zhongjiajie) [reply .github/workflows/code-analysys.yml]: I think it should be work https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution#example-only-run-job-for-specific-repository

Comment (CalvinKirs) [reply .github/workflows/code-analysys.yml]: It seems that his scope includes the submitted PR, I'm not sure, but I use another way.

**Review by zhongjiajie — APPROVED** [review]

File: tools/sonarcheck/check.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if [ ! "$SONAR_TOKEN" ]; then
+  echo "SONAR_TOKEN environment is null, skip check"
+  exit 0
+fi
+./mvnw --batch-mode verify sonar:sonar -Dmaven.test.skip=true -Dsonar.host.url=https://sonarcloud.io -Dsonar.organization=apache -Dsonar.projectKey=apache_incubator-seatunnel -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false -Dmaven.wagon.httpconnectionManager.ttlSeconds=120
```
Comment (kezhenxu94) [inline tools/sonarcheck/check.sh:21]: Hi @CalvinKirs , this basically says that, the check will only be executed in `push` (not `pull_request`) and those pull requests that are opened in this repo (not fork repo). Meaning the check only check committers' PRs 😂 IS that expected? I think all contributors might bring something that breaks Sonar Check

Comment (CalvinKirs) [reply tools/sonarcheck/check.sh:21]: It seems that for security reasons, secrets cannot be passed, here is the explanation of infra, am I wrong? 🤣

Comment (CalvinKirs) [reply tools/sonarcheck/check.sh:21]: https://issues.apache.org/jira/browse/INFRA-23101

Comment (kezhenxu94) [reply tools/sonarcheck/check.sh:21]: I know that secrets cannot be passed to forked repo. I mean do we have a "local mode" that, we can check and fail the CI, without uploading to Sonar cloud?

Comment (kezhenxu94) [reply tools/sonarcheck/check.sh:21]: Looks there is no local mode for Sonar check, let's keep it as is

**Pull request merged**